src/qibo/tests/test_hamiltonians.py

```diff
@@ -246,6 +248,38 @@ def test_hamiltonian_expectation_errors(backend):
         h.expectation("test")
 
 
+def test_hamiltonian_expectation_from_samples(backend):
+    """Test Hamiltonian expectation value calculation."""
+    obs0 = 2 * Z(0) * Z(1) + Z(0) * Z(2)
+    obs1 = 2 * Z(0) * Z(1) + Z(0) * Z(2) * I(3)
+    h0 = hamiltonians.SymbolicHamiltonian(obs0, backend=backend)
+    h1 = hamiltonians.SymbolicHamiltonian(obs1, backend=backend)
+    matrix = backend.to_numpy(h0.matrix)
+    c = Circuit(4)
+    c.add(gates.RX(0, np.random.rand()))
+    c.add(gates.RX(1, np.random.rand()))
+    c.add(gates.RX(2, np.random.rand()))
+    c.add(gates.RX(3, np.random.rand()))
+    c.add(gates.M(0, 1, 2, 3))
+    nshots = 10**5
+    result = c(nshots=nshots)
+    state = c()
+    freq = result.frequencies(binary=True)
+    Obs0 = hamiltonians.Hamiltonian(3, matrix).expectation_from_samples(
+        freq, qubit_map=None
+    )
+    Obs1 = h1.expectation(c().state())
```

Comment:
Here the `state` variable is not used anywhere. 

Also you don't need to execute the circuit twice to get the state, you can just do `result.state()` directly.

Reply:
Thanks for spotting this. Corrected.

Reply:
I think adding an aditional method to `CircuitResult`  is useful so I have implemented it.